multidict/_multidict_iter.c

```diff
@@ -0,0 +1,309 @@
+
+    return key;
+}
+
+static void
+multidict_iter_dealloc(MultidictIter *self)
+{
+    PyObject_GC_UnTrack(self);
+    Py_XDECREF(self->impl);
+    PyObject_GC_Del(self);
+}
+
+static int
+multidict_iter_traverse(MultidictIter *self, visitproc visit, void *arg)
+{
+    Py_VISIT(self->impl);
+    return 0;
+}
+
+static PyTypeObject multidict_items_iter_type = {
+    PyVarObject_HEAD_INIT(NULL, 0)
+    "multidict._multidict_iter._ItemsIter",         /* tp_name */
+    sizeof(MultidictIter),                          /* tp_basicsize */
+    0,                                              /* tp_itemsize */
+    (destructor)multidict_iter_dealloc,             /* tp_dealloc */
+    0,                                              /* tp_print */
+    0,                                              /* tp_getattr */
+    0,                                              /* tp_setattr */
+    0,                                              /* tp_reserved */
+    0,                                              /* tp_repr */
+    0,                                              /* tp_as_number */
+    0,                                              /* tp_as_sequence */
+    0,                                              /* tp_as_mapping */
+    0,                                              /* tp_hash */
+    0,                                              /* tp_call */
+    0,                                              /* tp_str */
+    0,                                              /* tp_getattro */
+    0,                                              /* tp_setattro */
+    0,                                              /* tp_as_buffer */
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,        /* tp_flags */
+    0,                                              /* tp_doc */
+    (traverseproc)multidict_iter_traverse,          /* tp_traverse */
+    0,                                              /* tp_clear */
+    0,                                              /* tp_richcompare */
+    0,                                              /* tp_weaklistoffset */
+    PyObject_SelfIter,                              /* tp_iter */
+    (iternextfunc)multidict_items_iter_iternext,    /* tp_iternext */
+    0,                                              /* tp_methods */
+    0,                                              /* tp_members */
+    0,                                              /* tp_getset */
+    0,                                              /* tp_base */
+    0,                                              /* tp_dict */
+    0,                                              /* tp_descr_get */
+    0,                                              /* tp_descr_set */
+    0,                                              /* tp_dictoffset */
+    0,                                              /* tp_init */
+    0,                                              /* tp_alloc */
+    (newfunc)multidict_iter_new,                    /* tp_new */
```

Comment:
Please take a look on `PyDictIterKey_Type` and family.
I think we don't need a `__new__` method but a C function to create an iterator.
Iterator instances should not be instantiated by user Python code but by views only.

Reply:
Ok. I added `__new__` just for compatibility with `_ItemsView` https://github.com/aio-libs/multidict/blob/master/multidict/_multidict.pyx#L497, etc.

And what we should do with `_ItemsView`, `_ValuesView`, `_KeysView`? Rewrite on C in this PR or separate?